api/pkg/apis/v1alpha1/providers/stage/materialize/materialize.go

```diff
@@ -127,46 +127,103 @@ func (i *MaterializeStageProvider) Process(ctx context.Context, mgrContext conte
 	for _, catalog := range catalogs {
 		for _, object := range prefixedNames {
 			if catalog.Spec.Name == object {
-				objectNamespace := "default"
-				if s, ok := inputs["objectNamespace"]; ok {
-					objectNamespace = s.(string)
-				}
 				objectData, _ := json.Marshal(catalog.Spec.Properties) //TODO: handle errors
 				name := catalog.Spec.Name
 				if s, ok := inputs["__origin"]; ok {
 					name = strings.TrimPrefix(catalog.Spec.Name, fmt.Sprintf("%s-", s))
 				}
 				switch catalog.Spec.Type {
 				case "instance":
-					err = utils.CreateInstance(ctx, i.Config.BaseUrl, name, i.Config.User, i.Config.Password, objectData, objectNamespace) //TODO: is using Spec.Name safe? Needs to support scopes
+					var instanceState model.InstanceState
+					err = json.Unmarshal(objectData, &instanceState)
+					if err != nil {
+						mLog.Errorf("Failed to unmarshal instance state for catalog %s: %s", name, err.Error())
+						return outputs, false, err
+					}
+					if instanceState.ObjectMeta.Name == "" {
```

Comment:
Can we extract the objectmetadata update part to a shared function

Reply:
fixed

Reply:
And I made the change to decide the name of materialized objects:
1. If inner object has name or display name in its spec, use it as the name
2. otherwise, use the name of wrapped catalog